File: contract/src/main/java/rrlane/leasing/contract/entity/LeasingContract.java
package rrlane.leasing.contract.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import lombok.Builder;
import lombok.Data;
import org.hibernate.annotations.GenericGenerator;
import rrlane.leasing.entity.Customer;
import rrlane.leasing.entity.Vehicle;

@Entity
@Table(name = "lcontract")
@Data
@Builder
public class LeasingContract {
    @Id
    @Column(name = "lcontract_id")
    @GeneratedValue(generator = "system-uuid")
    @GenericGenerator(name = "system-uuid", strategy = "uuid")
    private String id;
    @Column(name = "lcnum")
    private Integer contractNumber; //prop: source of truth is here
    @Column(name = "mrate")
    private Double monthlyRate; //prop: source of truth is here
    @OneToOne
    @JoinColumn(name = "vehicle_id") // one-to-one
    private Vehicle vehicle; // The vehicle in a contract may have its VIN as an empty String value
    @ManyToOne(optional = false)
    @JoinColumn(name = "customer_id")
    private Customer customer;

    public int hashCode() {
        return customer.getName().length();
    }

    public boolean equals(Object obj) {
        if (obj instanceof LeasingContract) {
            return this.id.equals(((LeasingContract) obj).getId());
        }
        return false;
    }

    public void setVehicle(Vehicle vehicle) {
        this.vehicle = vehicle;
        if (null == this.vehicle.getVin() || "".equals(this.vehicle.getVin())) this.vehicle.setVin("-");
    }
}

File: contractapi/src/main/kotlin/de/leasing/contract/api/entity/LeasingContract.kt
package de.leasing.contract.api.entity

import de.leasing.contract.api.entity.dto.LeasingContractReq
import de.leasing.contract.api.entity.dto.LeasingContractResp
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import jakarta.persistence.Temporal
import jakarta.persistence.TemporalType
import org.hibernate.annotations.UuidGenerator
import java.time.LocalDateTime

@Entity
@Table(name = "leasecontracts")
data class LeasingContract(
    @Column(name = "lcnum", length = 12) val contractNumber: String,
    @Column(name = "monthly_rate") val monthlyRate: Double,
    @Column(name = "vin", length = 12) val vehicleIN: String,
    @Column(name = "customer_name", length = 17) val customerName: String,
    @Column(name = "starts_at") @Temporal(TemporalType.TIMESTAMP) val leaseStartsAt: LocalDateTime? = null,
    @Column(name = "expires_at") @Temporal(TemporalType.TIMESTAMP) val leaseExpiresAt: LocalDateTime? = null
) {
    @Id
    @UuidGenerator
    lateinit var id: String

    fun toResp() = LeasingContractResp(
        id, contractNumber, monthlyRate, vehicleIN, customerName,leaseStartsAt,leaseExpiresAt
    )

    companion object {
        fun fromReq(req: LeasingContractReq): LeasingContract {
            return LeasingContract(
                req.contractNumber, req.monthlyRate, req.vehicleIN, req.customerName,req.leaseStartsAt,req.leaseExpiresAt
            )
        }
    }

}
